src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+
+    If the completion percentage is `None`, then the bar will render as
+    an indeterminate progress bar.
+    """
+    _refresh_timer: Timer | None
+    """Internal timer used to render the bar when it's in the indeterminate state."""
+
+    def watch__completion_percentage(self, percentage: float | None) -> None:
+        """Manage the timer that enables the indeterminate bar animation."""
+        if percentage is not None:
+            if self._refresh_timer:
+                self._refresh_timer.stop()
+                self._refresh_timer = None
+
+        else:
+            self._refresh_timer = self.set_interval(1 / 60, self.refresh)
+
+    def render(self) -> RenderResult:
+        """Render the bar with the correct portion filled."""
+        if self._completion_percentage is None:
+            return self.render_indeterminate()
+        else:
+            bar_style = (
+                self.get_component_rich_style("bar--bar")
+                if self._completion_percentage < 1
+                else self.get_component_rich_style("bar--complete")
+            )
+            return RenderableBar(
+                highlight_range=(0, self.size.width * self._completion_percentage),
+                highlight_style=Style.from_color(bar_style.color),
+                background_style=Style.from_color(bar_style.bgcolor),
+            )
+
+    def render_indeterminate(self) -> RenderResult:
+        """Render a frame of the indeterminate progress bar animation."""
+        width = self.size.width
+        highlighted_bar_width = 0.25 * width
+        # Width used to enable the visual effect of the bar going into the corners.
+        total_imaginary_width = width + highlighted_bar_width
+
+        speed = 30  # Cells per second.
+        # Compute the position of the bar.
+        start = (speed * monotonic()) % (2 * total_imaginary_width)
+        if start > total_imaginary_width:
+            # If the bar is to the right of its width, wrap it back from right to left.
+            start = 2 * total_imaginary_width - start  # = (tiw - (start - tiw))
+        start -= highlighted_bar_width
+        end = start + highlighted_bar_width
+
+        bar_style = self.get_component_rich_style("bar--indeterminate")
+        return RenderableBar(
+            highlight_range=(max(0, start), min(end, width)),
+            highlight_style=Style.from_color(bar_style.color),
+            background_style=Style.from_color(bar_style.bgcolor),
+        )
+
+
+class PercentageStatus(Label):
+    """A label to display the percentage status of the progress bar."""
+
+    _completion_percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed or `None` if indeterminate."""
+
+    DEFAULT_CSS = """
+    PercentageStatus {
+        width: 5;
+        content-align-horizontal: right;
+    }
+    """
+
+    def watch__completion_percentage(self, percentage: float | None) -> None:
```

Comment:
Setting the renderable won't prompt a refresh. There is an `update` method which will set the renderable and prompt a refresh.

We want to avoid unnecessary updates. If the rounded percentage hasn't changed, we don't want to refresh.

Suggest we have a second reactive for the status text, which refresh when it changes.